### ISSUE TITLE
Allow Views to be grouped in Admin Menu

### ISSUE THREAD

**Issue by yorkshire-pudding** [backdrop/backdrop-issues#5895]
## Description of the need

As a site builder on a site with a lot of views
I want to be able to navigate easily to all views via the admin menu
So that I can quickly access the view for editing

When there are a lot of views, then the admin menu goes off the bottom of the page and the lower links can't be used.
Core adds some 3 views and the Settings, webform adds 4 views, other modules add views, then the site architect adds a few views and soon you have to go to the listing to find the view; you can't do it from the menu.

![image](https://user-images.githubusercontent.com/20357797/209148471-c99cb716-899f-4d10-93f3-d299b2270654.png)

## Proposed solution

- User (and modules) can add a single tag to each view (not display level) via config and the UI. This could be a taxonomy tag or it could be a very simple word tag with no wider application.
- User access `admin/structure/views` and all tags form a subfolder within the menu
- Listing of views allows filtering by tag.

## Alternatives that have been considered

Investigated the view api to see if could alter things that way
Investigated the menu links to see if could manually create groups but the view listing menu items are dynamically generated.

Is there a Drupal or Backdrop contributed module that accomplishes this?
...if so, then please provide a link.

I don't think so
## Additional information

Add any other information or screenshots that would help.

![image](https://user-images.githubusercontent.com/20357797/209151060-ea0f991e-7d08-4571-a952-7173637b3063.png)

## Draft of feature description for Press Release (1 paragraph at most)

Backdrop now includes the ability to group views by a tag for easy sorting within the administration menu and easy filtering on the 'List views' page.  Tags are set per view and can be set both by modules creating views and by users creating and editing views.

**Comment by bugfolder** [comment]
I run into this issue too and so would love to see a solution to the problem.

All Views currently have a "View tags" field, which is a comma-delimited list. The help text for this field says "Optionally, enter a comma delimited list of tags for this view to use in filtering and sorting views on the administrative page." This info is displayed on the administrative page, but there's no apparent ability to sort them by these tags. At any rate, since multiple tags are possible in that field, using the existing tag field to define an admin menu hierarchy might be challenging. So perhaps a new field ("Group"?) might be in order.

**Comment by yorkshire-pudding** [comment]
> All Views currently have a "View tags" field, which is a comma-delimited list.

I hadn't spotted that before; you're right that despite the promises, it doesn't do anything for filtering or sorting.  And I agree that grouping a multi value list could be difficult in the admin menu.

**Comment by indigoxela** [comment]
I think, some on-page filtering like the module listing page should be relatively easy. And very helpful.

**Comment by argiepiano** [comment]
Yes, it'd be great to have at least some of the D7 Views UI filtering capabilities.  https://forum.backdropcms.org/forum/views-administrative-page-doesnt-show-filters

**Comment by indigoxela** [comment]
> ... it'd be great to have at least some of the D7 Views UI filtering capabilities...

There aren't any filter options in D7, either. Only sorting by tag/path is possible, which is sparse.

**Comment by bugfolder** [comment]
> There aren't any filter options in D7, either.

Oh yes there are, but they need to be enabled in the Views settings.

![d7_list_views](https://user-images.githubusercontent.com/29214084/209178652-38505787-7885-4141-9606-ee73d7928ade.jpg)

That's a different issue than this one, so I've opened an issue specifically for it.

https://github.com/backdrop/backdrop-issues/issues/5896

**Comment by indigoxela** [comment]
Oh, right, it's a setting, that's not available in B. :+1:

**Comment by klonos** [comment]
Cross-linking some relevant issues: #2074 / #1897 / #503

Special mention to  #2809 and #520, since [smartmenus](https://www.smartmenus.org) have this "scrolling menus" feature, which would solve the problem reported by @yorkshire-pudding here:
![smartmenus-long_menus_within_viewport](https://cloud.githubusercontent.com/assets/2423362/18275209/53f4d808-7489-11e6-985c-6ceb4b96d729.gif)

**Comment by yorkshire-pudding** [comment]
Thanks @klonos - I think switching to smart menus would be good and would fix the immediate symptom that prompted this request, though I still think there is a use case for being able to group views under categories to make it easier for the site admin to find.

**Comment by klonos** [comment]
Fair enough @yorkshire-pudding ...looking at the b.org main site, I see that we have the same problem there with layouts and content types. So sites of similar complexity are likely to need that menu item grouping feature for other things as well. Would be nice to figure out a way to allow it to be applied to multiple menu "sections" if possible.

**Comment by yorkshire-pudding** [comment]
@klonos - yes, I could well see content types and paragraph types easily getting to the size where grouping would greatly improve the usability for site architects.

**Comment by yorkshire-pudding** [comment]
Discussed in the UX/Design meeting.  Recommended that this isn't done in core initially.
We should look first at whether we can use the `admin_bar` module API to alter the menu.
Also discussed using the existing tags field and thought it may not be a problem to have multiple tags here as it would then appear in each virtual folder.
Important not to change paths of the views.

**Comment by bugfolder** [comment]
So...big thanks to @jenlampton for suggesting how to go about this. Here's a contrib module that does this. For any View that has tags, there is a submenu for that tag containing all Views with that tag (so a View with multiple tags will show up in multiple submenus).

https://github.com/backdrop-contrib/admin_bar_views_grouping

Suggestions for improvements (or things I overlooked!) are of course welcome.

The changes seem modest enough that this could perhaps be added to core `admin_bar` module, with a controlling checkbox added to the Administration bar settings (or perhaps Views settings, via hook_form_alter()).

**Comment by yorkshire-pudding** [comment]
This is amazing; thank you so much @bugfolder !!  It works a treat.
Two very minor suggestions:
- Add in the Readme a mention that tags can be set and changed in each view by following the "Change View Name/Description" link at the top; I wasn't aware that these tags existed until it was mentioned on this issue.
- Perhaps the tags should be represented in sentence case within the menu, so using `ucfirst()` around the `check_plain($tag)` on [line 39](https://github.com/backdrop-contrib/admin_bar_views_grouping/blob/11c69de932337652f969ec62e0d59df2b5c0a868/admin_bar_views_grouping.module#L39); I'm not 100% certain this is right, but the menu looked odd to me with lowercase folder headings, even though that is an accurate representation of the tag.

**Comment by bugfolder** [comment]
> Add in the Readme a mention that tags can be set and changed...

Done!

> Perhaps the tags should be represented in sentence case within the menu

This one I'm leaning against, but am open to further persuasion. I agree with the oddness of appearance. On the other hand:
* Views-builders always have the option to make their tags sentence case;
* we don't force their View names to be sentence case, so if they've named their View with a lower-case name, it will show up as lower-case as well.
* someone might want the contrast between tags and View names so that the submenu items look different from the Views items at the same level (for Views that don't have tags);

I'd also welcome further input on whether this could/should be absorbed into the core `admin_bar` module (with a checkbox in the settings to enable/disable it) or should stay contrib.

**Comment by yorkshire-pudding** [comment]
You make a good case for leaving the case of the tags intact.
I think there is a strong case for including this in core; views tags are already in core (and currently unused) and this makes use of them to make it easy to access all views from the admin bar with its default settings.  At present, it feels like the admin menu is broken when the view menu links go off the page.

Of course, @klonos made the good point that this can also apply to layouts and content types; these currently have no tag field so any work to improve these would be a separate case.

**Comment by bugfolder** [comment]
The newly added Permissions filter added support for filtered pages via a query parameter on the URL, e.g., http://mysite.com/admin/config/people/permissions?search=admin.

https://github.com/backdrop/backdrop-issues/issues/5896 is an issue to add similar instant filtering to Views. If that's done the same way, then it should be possible to create URLs to filtered-by-tag Views listings pages also via query parameters.

And then when that is in place, we can make the tag menu items link to the filtered Views listing (currently they link to the full Views listing).

**Comment by yorkshire-pudding** [comment]
@bugfolder - are you willing to release this as contrib as it could be a while before it is in core, if it is agreed to put it in?

**Comment by bugfolder** [comment]
> are you willing to release this as contrib?

Yes. I thought I'd give it a little time to get feedback on the UI and approach (and thanks for yours!), but it seems like it's stable enough to go ahead and release.

And if tag-specific Views listing URLs become possible, it should be easy to add that to the module backward-compatibly.

And then maybe this issue becomes "should this contrib module be merged into core?"

**Comment by bugfolder** [comment]
https://backdropcms.org/project/admin_bar_views_grouping

**Comment by klonos** [comment]
Thanks @bugfolder 🙏🏼 ...I just added the first feature request: https://github.com/backdrop-contrib/admin_bar_views_grouping/issues/1 😉
